### PR TITLE
Implement propagate_const in the framework

### PR DESCRIPTION
**Pull request by wmtan**
This PR supersedes PR #13015.
This PR uses the propagate_const template for most pointers in the framework that are data members of classes or structs, and makes the knock on changes caused by this. This means that a const member function of the class will no longer be able to modify the data to which the pointer points. This PR is therefore a partial implementation of issue #12840. The member pointers in the framework that are not modified by this PR are:
1) Those in FWLite (will be done in a separate PR).
2) Pointers to const or pointers already declared mutable (no need to modify).
3) Those pointers whose modifications require changes that were too extensive or complex to be done in this PR. These will be dealt within a case by case basis.
Other minor changes made in a very few places in affected code:
a) a few boost::scoped_ptr changed to std::unique_ptr
b) a few std::auto_ptr changed to std::unique_ptr
c) iterator loops replaced by range based for loops
d) nullptr replaces 0 for pointers

This PR does not change any instances of boost::shared_ptr to std::shared_ptr, because most of these have already been changed in previous PR's, and the ones remaining are not trivial to change. These will be changed in subsequent PR's where feasible.

### CODE REVIEW

**Comment by wmtan** [comment]
@cmsbuild please test

File: FWCore/Framework/interface/InputSource.h [hunk 1]
```diff
@@ -414,7 +417,7 @@ namespace edm {
     virtual void endRun(Run&);
     virtual void beginJob();
     virtual void endJob();
-    virtual SharedResourcesAcquirer* resourceSharedWithDelayedReader_() const;
+    virtual SharedResourcesAcquirer* resourceSharedWithDelayedReader_();
```
Comment (wmtan) [inline FWCore/Framework/interface/InputSource.h]: OK to return ptr to non-const in const function because the SharedResourcesAcquirer is mutable.

File: FWCore/Framework/interface/InputSource.h [hunk 2]
```diff
@@ -184,7 +189,7 @@ namespace edm {
     }
     
     /// Returns nullptr if no resource shared between the Source and a DelayedReader
-    SharedResourcesAcquirer* resourceSharedWithDelayedReader() const;
```
Comment (wmtan) [inline FWCore/Framework/interface/InputSource.h]: OK to return pointer to non-const in const function because SharedResourcesAcquire is mutable

File: DataFormats/Common/interface/ProductData.h [hunk 3]
```diff
@@ -9,7 +9,7 @@ is the storage unit of such information.
 ----------------------------------------------------------------------*/
 
 #include "DataFormats/Provenance/interface/Provenance.h"
-#include "FWCore/Utilities/interface/propagate_const.h"
+#include "FWCore/Utilities/interface/propagate_const_safe.h"
```
Comment (Dr15Jones) [inline DataFormats/Common/interface/ProductData.h]: This include nor the one it replaced appears to be needed.

File: DataFormats/Provenance/interface/IndexIntoFile.h [hunk 4]
```diff
@@ -1042,7 +1043,7 @@ namespace edm {
       void fillRunOrLumiIndexes() const;
 
       void fillUnsortedEventNumbers() const;
-      void resetEventFinder() const {transient_.eventFinder_.reset();}
+      void resetEventFinder() const {transient_.eventFinder_ = nullptr;} // propagate_const<T> has no reset() function
```
Comment (Dr15Jones) [inline DataFormats/Provenance/interface/IndexIntoFile.h]: Seems like we should investigate why ```transient_``` is mutable.

File: DataFormats/Provenance/interface/ProductHolderIndexHelper.h [hunk 5]
```diff
@@ -54,6 +54,7 @@ ProductRegistry is frozen.
 #include "FWCore/Utilities/interface/ProductHolderIndex.h"
 #include "FWCore/Utilities/interface/ProductKindOfType.h"
 #include "FWCore/Utilities/interface/TypeID.h"
+#include "FWCore/Utilities/interface/propagate_const_safe.h"
```
Comment (Dr15Jones) [inline DataFormats/Provenance/interface/ProductHolderIndexHelper.h]: I would have prefered to just depend on ```propagate_const.h``` and only include ```propagate_const_safe``` in the cases where the stand alone function is used.

File: FWCore/Framework/interface/InputSource.h [hunk 6]
```diff
@@ -424,7 +427,7 @@ namespace edm {
 
   private:
 
-    std::shared_ptr<ActivityRegistry> actReg_;
+    mutable std::shared_ptr<ActivityRegistry> actReg_;
```
Comment (Dr15Jones) [inline FWCore/Framework/interface/InputSource.h]: Why this change to ```mutable```? is the ```shared_ptr``` now changed in a ```const``` function?

Comment (wmtan) [inline FWCore/Framework/interface/InputSource.h]: @Dr15Jones Very likely yes. Also, I asked you if the ActivityRegistry could be mutable, and you said yes, because it is thread safe. I know for sure that some of the actReg_ data members needed to be modified by a const function in one or more classes, but I am not 100% sure that this is true in each and every class that holds the shared pointer.

File: DataFormats/Common/interface/ProductData.h (hunk 3 above)
Comment (wmtan) [inline DataFormats/Common/interface/ProductData.h]: @Dr15Jones I'll remove it if not needed, which is very likely correct.

File: FWCore/Framework/interface/InputSourceDescription.h [hunk 7]
```diff
@@ -51,7 +51,7 @@ namespace edm {
     std::shared_ptr<ProductRegistry> productRegistry_;
     std::shared_ptr<BranchIDListHelper> branchIDListHelper_;
     std::shared_ptr<ThinnedAssociationsHelper> thinnedAssociationsHelper_;
-    std::shared_ptr<ActivityRegistry> actReg_;
+    mutable std::shared_ptr<ActivityRegistry> actReg_;
```
Comment (Dr15Jones) [inline FWCore/Framework/interface/InputSourceDescription.h]: Here's another

File: FWCore/Framework/interface/PrincipalGetAdapter.h [hunk 8]
```diff
@@ -235,7 +236,7 @@ namespace edm {
     ModuleDescription const& md_;
     
     EDConsumerBase const* consumer_;
-    SharedResourcesAcquirer* resourcesAcquirer_;
+    mutable SharedResourcesAcquirer* resourcesAcquirer_;
```
Comment (Dr15Jones) [inline FWCore/Framework/interface/PrincipalGetAdapter.h]: This also changes the semantics. I don't think there was a need for ```mutable```.

File: FWCore/Framework/interface/ScheduleItems.h [hunk 9]
```diff
@@ -53,12 +54,21 @@ namespace edm {
...
+    std::shared_ptr<SignallingProductRegistry const> preg() const {return get_underlying_safe(preg_);}
+    std::shared_ptr<SignallingProductRegistry>& preg() {return get_underlying_safe(preg_);}
+    std::shared_ptr<BranchIDListHelper const> branchIDListHelper() const {return get_underlying_safe(branchIDListHelper_);}
+    std::shared_ptr<BranchIDListHelper>& branchIDListHelper() {return get_underlying_safe(branchIDListHelper_);}
+    std::shared_ptr<ThinnedAssociationsHelper const> thinnedAssociationsHelper() const {return get_underlying_safe(thinnedAssociationsHelper_);}
+    std::shared_ptr<ThinnedAssociationsHelper>& thinnedAssociationsHelper() {return get_underlying_safe(thinnedAssociationsHelper_);}
+    std::shared_ptr<ProcessConfiguration const> processConfiguration() const {return get_underlying_safe(processConfiguration_);}
+    std::shared_ptr<ProcessConfiguration>& processConfiguration() {return get_underlying_safe(processConfiguration_);}
+
+    mutable std::shared_ptr<ActivityRegistry> actReg_;
```
Comment (Dr15Jones) [inline FWCore/Framework/interface/ScheduleItems.h]: another instance

File: FWCore/Framework/src/EventSetupRecordProvider.cc [hunk 10]
```diff
@@ -288,24 +286,24 @@ EventSetupRecordProvider::proxyProviderDescriptions() const
 }
 
 boost::shared_ptr<DataProxyProvider> 
-EventSetupRecordProvider::proxyProvider(const ComponentDescription& iDesc) const {
+EventSetupRecordProvider::proxyProvider(ComponentDescription const& iDesc) {
    using std::placeholders::_1;
-   std::vector<boost::shared_ptr<DataProxyProvider> >::const_iterator itFound =
-   std::find_if(providers_.begin(),providers_.end(),
+   //std::vector<boost::shared_ptr<DataProxyProvider> >::const_iterator itFound =
```
Comment (Dr15Jones) [inline FWCore/Framework/src/EventSetupRecordProvider.cc]: please remove commented out code.

File: FWCore/Framework/src/Schedule.cc [hunk 11]
```diff
@@ -927,8 +940,8 @@ namespace edm {
     
     globalSchedule_->replaceModule(newMod,iLabel);
 
-    for(auto s: streamSchedules_) {
-      s->replaceModule(newMod,iLabel);
+    for(auto& s: streamSchedules_) {
+      get_underlying_safe(s)->replaceModule(newMod,iLabel);
```
Comment (Dr15Jones) [inline FWCore/Framework/src/Schedule.cc]: Why the need for ```get_underlying_safe```? I would have thought ``operator->``` would be enough.

File: FWCore/Framework/src/Schedule.cc [hunk 12]
```diff
@@ -1052,7 +1065,7 @@ namespace edm {
   }
                           
   void
-  Schedule::getTriggerTimingReport(TriggerTimingReport& rep) const {
+  Schedule::getTriggerTimingReport(TriggerTimingReport& rep) {
```
Comment (Dr15Jones) [inline FWCore/Framework/src/Schedule.cc]: Why is this non-const? Is there something wrong with the ```const```ness of ```summaryTimeKeeper_```?

File: FWCore/MessageLogger/interface/MessageSender.h [hunk 13]
```diff
@@ -4,6 +4,7 @@
 #include "FWCore/MessageLogger/interface/ELstring.h"
 #include "FWCore/MessageLogger/interface/ELseverityLevel.h"
 #include "FWCore/MessageLogger/interface/ErrorObj.h"
+#include "FWCore/Utilities/interface/propagate_const_safe.h"
```
Comment (Dr15Jones) [inline FWCore/MessageLogger/interface/MessageSender.h]: Is this needed?

File: FWCore/Framework/src/EventSetupRecordProvider.cc (hunk 10 above)
Comment (wmtan) [inline FWCore/Framework/src/EventSetupRecordProvider.cc]: Already done in my working area.

File: FWCore/Modules/src/MulticoreRunLumiEventChecker.cc [hunk 14]
```diff
@@ -78,7 +79,7 @@ class MulticoreRunLumiEventChecker : public edm::EDAnalyzer {
    unsigned int numberOfEventsLeftBeforeSearch_;
    bool mustSearch_;
    
-   std::shared_ptr<boost::thread> listenerThread_;
+   edm::propagate_const<std::shared_ptr<boost::thread>> listenerThread_;
```
Comment (Dr15Jones) [inline FWCore/Modules/src/MulticoreRunLumiEventChecker.cc]: NOTE TO SELF: we should switch to ```std::thread```

File: FWCore/Sources/interface/VectorInputSourceDescription.h [hunk 15]
```diff
@@ -8,6 +8,7 @@ a VectorinputSource that does not come in through the ParameterSet
 ----------------------------------------------------------------------*/
 
 #include "FWCore/Framework/src/PreallocationConfiguration.h"
+#include "FWCore/Utilities/interface/propagate_const_safe.h"
```
Comment (Dr15Jones) [inline FWCore/Sources/interface/VectorInputSourceDescription.h]: Is this needed?

File: FWCore/Utilities/interface/RandomNumberGenerator.h [hunk 16]
```diff
@@ -153,10 +153,10 @@ namespace edm {
     /// These are the only functions most modules should call.
 
     /// Use this engine in event methods
-    virtual CLHEP::HepRandomEngine& getEngine(StreamID const&) const = 0;
```
Comment (Dr15Jones) [inline FWCore/Utilities/interface/RandomNumberGenerator.h]: What lead to this being non-const?

File: FWCore/Utilities/interface/propagate_const_safe.h [hunk 17]
```diff
@@ -0,0 +1,59 @@
+#ifndef FWCore_Utilities_propagate_const_safe_h
```
Comment (Dr15Jones) [inline FWCore/Utilities/interface/propagate_const_safe.h]: This should be renamed to ```get_underlying_safe.h```.

File: IOMC/RandomEngine/src/RandomNumberGeneratorService.cc [hunk 18]
```diff
@@ -215,11 +215,11 @@ namespace edm {
       }
       unsigned int moduleID = mcc->moduleDescription()->id();
 
-      std::vector<ModuleIDToEngine> const& moduleIDVector = streamModuleIDToEngine_.at(streamID.value());
+      std::vector<ModuleIDToEngine>& moduleIDVector = streamModuleIDToEngine_.at(streamID.value());
```
Comment (Dr15Jones) [inline IOMC/RandomEngine/src/RandomNumberGeneratorService.cc]: This changed the semantics. Previously we stated this couldn't change, now it does.

Comment (Dr15Jones) [inline IOMC/RandomEngine/src/RandomNumberGeneratorService.cc]: Presumably this was caused by the ```ModuleIDToEngine``` needing to return a non-const random number generator engine.

File: IOPool/Input/src/RootFile.cc [hunk 19]
```diff
@@ -638,7 +642,7 @@ namespace edm {
   }
 
   std::unique_ptr<FileBlock>
-  RootFile::createFileBlock() const {
+  RootFile::createFileBlock() {
```
Comment (Dr15Jones) [inline IOPool/Input/src/RootFile.cc]: What part of the ```FileBlock``` constructor requires this to be non-const?

File: IOPool/Output/src/RootOutputTree.cc [hunk 20]
```diff
@@ -181,8 +178,8 @@ namespace edm {
     if(inputTree == nullptr) return false;
 
     // Do the sub-branches match in the input and output. Extra sub-branches in the input are OK for fast cloning, but not in the output.
-    for(std::vector<TBranch*>::const_iterator it = readBranches_.begin(), itEnd = readBranches_.end(); it != itEnd; ++it) {
-      TBranchElement* outputBranch = dynamic_cast<TBranchElement*>(*it);
+    for(auto const& outputBr : readBranches_) {
+      TBranchElement* outputBranch = dynamic_cast<TBranchElement*>(outputBr);
```
Comment (Dr15Jones) [inline IOPool/Output/src/RootOutputTree.cc]: Did that cast away const?

**Comment by Dr15Jones** [comment]
I've finished reviewing the code. Now I need a stiff drink :D.

File: DataFormats/Provenance/interface/ProductHolderIndexHelper.h (hunk 5 above)
Comment (wmtan) [inline DataFormats/Provenance/interface/ProductHolderIndexHelper.h]: @Dr15Jones This will be done in my next commit

File: FWCore/Framework/interface/InputSourceDescription.h (hunk 7 above)
Comment (wmtan) [inline FWCore/Framework/interface/InputSourceDescription.h]: @Dr15Jones Reverted in next commit.

File: FWCore/Framework/interface/PrincipalGetAdapter.h (hunk 8 above)
Comment (wmtan) [inline FWCore/Framework/interface/PrincipalGetAdapter.h]: @Dr15Jones The mutability of pointers to the SharedResourcesAcquirer will be reverted in my next commit.

File: FWCore/Framework/interface/ScheduleItems.h (hunk 9 above)
Comment (wmtan) [inline FWCore/Framework/interface/ScheduleItems.h]: @Dr15Jones Also reverted in next commit

File: FWCore/Framework/src/Schedule.cc (hunk 11 above)
Comment (wmtan) [inline FWCore/Framework/src/Schedule.cc]: @Dr15Jones You are correct. get_underlying_safe here will be removed in the next commit.

File: FWCore/Framework/src/Schedule.cc (hunk 12 above)
Comment (wmtan) [inline FWCore/Framework/src/Schedule.cc]: @Dr15Jones getTriggerTimingReport actually fills in the timing report, so summaryTimeKeeper is not const. This was revealed by propagate_const. I will investigate how much work it would be to make summaryTimeKeeper const, but it would be outside the scope of implementing propagate_const.

File: FWCore/MessageLogger/interface/MessageSender.h (hunk 13 above)
Comment (wmtan) [inline FWCore/MessageLogger/interface/MessageSender.h]: @Dr15Jones No. Removed in next commit.

File: FWCore/Sources/interface/VectorInputSourceDescription.h (hunk 15 above)
Comment (wmtan) [inline FWCore/Sources/interface/VectorInputSourceDescription.h]: @Dr15Jones No. Removed in next commit.

File: FWCore/Utilities/interface/propagate_const_safe.h (hunk 17 above)
Comment (wmtan) [inline FWCore/Utilities/interface/propagate_const_safe.h]: @Dr15Jones Done in next commit

File: IOPool/Output/src/RootOutputTree.cc (hunk 20 above)
Comment (wmtan) [inline IOPool/Output/src/RootOutputTree.cc]: @Dr15Jones Not really. In the C++ meaning of constness, which does not propagate, this cast cannot cast away constness. In the propagating meaning of constness, this was not const in the first place because propagate_const is not implemented yet for readBranches_. It was one of the "hard cases."

File: IOPool/Input/src/RootFile.cc (hunk 19 above)
Comment (wmtan) [inline IOPool/Input/src/RootFile.cc]: @Dr15Jones Good catch! The FileBlock contains a shared_ptr&lt;BranchChildren&gt;. By changing it to shared_ptr&lt;BranchChildren const&gt;, I was able to make createFileBlock() const again. This will be in the next commit.

File: IOMC/RandomEngine/src/RandomNumberGeneratorService.cc (hunk 18 above)
Comment (wmtan) [inline IOMC/RandomEngine/src/RandomNumberGeneratorService.cc]: @Dr15Jones Yes, it was caused by the ModuleIDToEngine needing to return a non-const random number generator engine.

File: FWCore/Utilities/interface/RandomNumberGenerator.h (hunk 16 above)
Comment (wmtan) [inline FWCore/Utilities/interface/RandomNumberGenerator.h]: @Dr15Jones IT was caused by the ModuleIDToEngine needing to return a non-const random number generator engine.

**Comment by wmtan** [comment]
@cmsbuild The clang build error occurs in the base IB without this PR, and is therefore not caused by this PR.

File: FWCore/Framework/src/Schedule.cc (hunk 12 above)
Comment (Dr15Jones) [inline FWCore/Framework/src/Schedule.cc]: @wmtan I looked at this can remain ```const``` if we simply change ```SystemTimKeeper::fillTriggerTimingReport``` to be a ```const``` function. I already tested changing that member function to be ```const``` and everything compiled.

**Comment by Dr15Jones** [comment]
Once the change to ```SystemTimeKeeper``` is done I will approve the change.

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
+1

**Comment by wmtan** [comment]
@cmsbuild @Dr15Jones @davidlange6 The build error has nothing to do with this PR. Your disk ran out of space.

**Comment by wmtan** [comment]
@cmsbuild please test
